ext/phar/phar.c

```diff
@@ -600,49 +570,161 @@ int phar_open_parsed_phar(char *fname, size_t fname_len, char *alias, size_t ali
 /* }}}*/
 
 /**
- * Parse out metadata from the manifest for a single file
- *
- * Meta-data is in this format:
- * [len32][data...]
- *
- * data is the serialized zval
+ * Attempt to serialize the data.
+ * Callers are responsible for handling EG(exception) if one occurs.
  */
-int phar_parse_metadata(char **buffer, zval *metadata, uint32_t zip_metadata_len) /* {{{ */
+void phar_metadata_tracker_try_ensure_has_serialized_data(phar_metadata_tracker *tracker, int persistent) /* {{{ */
 {
-	php_unserialize_data_t var_hash;
+	php_serialize_data_t metadata_hash;
+	smart_str metadata_str = {0};
+	if (tracker->str || Z_ISUNDEF(tracker->val)) {
+		/* Already has serialized the value or there is no value */
+		return;
+	}
+	/* Assert it should not be possible to create raw zvals in a persistent phar (i.e. from cache_list) */
+	ZEND_ASSERT(!persistent);
 
-	if (zip_metadata_len) {
-		const unsigned char *p;
-		unsigned char *p_buff = (unsigned char *)estrndup(*buffer, zip_metadata_len);
-		p = p_buff;
+	PHP_VAR_SERIALIZE_INIT(metadata_hash);
+	php_var_serialize(&metadata_str, &tracker->val, &metadata_hash);
+	PHP_VAR_SERIALIZE_DESTROY(metadata_hash);
+	if (!metadata_str.s) {
+		return;
+	}
+	if (persistent) {
+		tracker->str = zend_string_init(ZSTR_VAL(metadata_str.s), ZSTR_LEN(metadata_str.s), persistent);
+		zend_string_release(metadata_str.s);
+	} else {
+		tracker->str = metadata_str.s;
+	}
+}
+/* }}} */
+
+/**
+ * Parse out metadata when phar_metadata_tracker_has_data is true
+ */
+int phar_metadata_tracker_unserialize_or_copy(phar_metadata_tracker *tracker, zval *metadata, int persistent, zval *unserialize_options, const char* method_name) /* {{{ */
+{
+	const zend_bool has_unserialize_options = unserialize_options != NULL && Z_TYPE_P(unserialize_options) == IS_ARRAY &&
+		zend_array_count(Z_ARRVAL_P(unserialize_options)) > 0;
+	if (EG(exception)) {
+		return FAILURE;
+	}
+
+	if (Z_ISUNDEF(tracker->val) || persistent || has_unserialize_options) {
+		const char *start;
+		/* Assert it should not be possible to create raw data in a persistent phar (i.e. from cache_list) */
+
+		ZEND_ASSERT(has_unserialize_options || Z_ISUNDEF(tracker->val));
+		if (has_unserialize_options && tracker->str == NULL) {
```

Comment:
So, is the idea here that we can call getMetadata multiple times with different unserialize options and because of that have to go through an artificial serialize + unserialize cycle here?

Reply:
I feel like this is expending unnecessary effort and complexity for something nobody is going to care about. I would just spec the function to say "returns the metadata if already unserialized, or unserializes it with the given options if not".

Reply:
> So, is the idea here that we can call getMetadata multiple times with different unserialize options and because of that have to go through an artificial serialize + unserialize cycle here?

Only if there wasn't serialized data. But in the latest commit, I realize this is unnecessary - both setMetaData and reading a phar file ensure there is serialized data.

Reply:
I did that a few days ago but forgot to push that.

I also added another commit

```
    Improve robustness of `Phar*->setMetadata()`
    
    Add sanity checks for edge cases freeing metadata, when destructors
    or serializers modify the phar recursively.
    
    Typical use cases of php have read-only phars (from the ini setting) and would not be affected.
```